graph/number_of_paths.cpp

```diff
@@ -0,0 +1,131 @@
+                                  std::uint32_t u, 
+                                  std::uint32_t v, 
+                                  std::uint32_t n, 
+                                  std::vector<bool>& visited) {
+        if (u == v) {
+            return 1; // Base case: Reached the destination node
+        }
+
+        visited[u] = true; // Mark the current node as visited
+        std::uint32_t path_count = 0; // Count of all paths from `u` to `v`
+
+        for (std::uint32_t i = 0; i < n; i++) {
+            if (A[u][i] == 1 && !visited[i]) { // Check if there is an edge and the node is not visited
+                path_count += count_paths_dfs(A, i, v, n, visited); // Recursively explore paths from `i` to `v`
+            }
+        }
+
+        visited[u] = false; // Unmark the current node as visited (backtracking)
+        return path_count;
+    }
+
+    /**
+     * @brief Counts the number of paths from node `u` to node `v` in a directed graph
+     * using Depth First Search (DFS)
+     * 
+     * @param A adjacency matrix representing the graph (1: edge exists, 0: no edge)
+     * @param u the starting node
+     * @param v the destination node
+     * @param n the number of nodes in the graph
+     * @returns the number of paths from node `u` to node `v`
+     */
+    std::uint32_t count_paths(const std::vector<std::vector<std::uint32_t>>& A, 
+                              std::uint32_t u, 
+                              std::uint32_t v, 
+                              std::uint32_t n) {
+        std::vector<bool> visited(n, false); // Initialize a visited vector for tracking nodes
+        return count_paths_dfs(A, u, v, n, visited);
+    }
+
+} // namespace graph
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // Test case 1: Simple directed graph with multiple paths
+    std::vector<std::vector<std::uint32_t>> graph1 = {
+        {0, 1, 0, 1, 0}, 
+        {0, 0, 1, 0, 1}, 
+        {0, 0, 0, 0, 1}, 
+        {0, 0, 1, 0, 0}, 
+        {0, 0, 0, 0, 0}  
+    };
+    std::uint32_t n1 = 5, u1 = 0, v1 = 4;
+    assert(graph::count_paths(graph1, u1, v1, n1) == 3);  // There are 3 paths from node 0 to 4
+
+    // Test case 2: No possible path (disconnected graph)
+    std::vector<std::vector<std::uint32_t>> graph2 = {
+        {0, 1, 0, 0, 0}, 
+        {0, 0, 0, 0, 0}, 
+        {0, 0, 0, 0, 1}, 
+        {0, 0, 1, 0, 0}, 
+        {0, 0, 0, 0, 0}  
+    };
+    std::uint32_t n2 = 5, u2 = 0, v2 = 4;
+    assert(graph::count_paths(graph2, u2, v2, n2) == 0);  // No path from node 0 to 4
+
+    // Test case 3: Cyclic graph with multiple paths
+    std::vector<std::vector<std::uint32_t>> graph3 = {
+        {0, 1, 0, 0, 0}, 
+        {0, 0, 1, 1, 0}, 
+        {1, 0, 0, 0, 1}, 
+        {0, 0, 1, 0, 1}, 
+        {0, 0, 0, 0, 0}  
+    };
+    std::uint32_t n3 = 5, u3 = 0, v3 = 4;
+    assert(graph::count_paths(graph3, u3, v3, n3) == 3);  // There are 3 paths from node 0 to 4
+
```

Comment:
add a test case for empty array
std::vector<std::vector<std::uint32_t>> graph5 = {{}};

this should return 0 as there is no path

Reply:
I've made the latest changes as per your feedback. I hope this meets the requirements now. Looking forward to your review.